iroh-relay/src/protos/handshake.rs

```diff
@@ -0,0 +1,750 @@
+/// allowed to happen at any time there.
+trait Frame {
+    /// The frame type this frame is identified by and prefixed with
+    const TAG: FrameType;
+}
+
+impl<T: Frame> Frame for &T {
+    const TAG: FrameType = T::TAG;
+}
+
+impl Frame for ServerChallenge {
+    const TAG: FrameType = FrameType::ServerChallenge;
+}
+
+impl Frame for ClientAuth {
+    const TAG: FrameType = FrameType::ClientAuth;
+}
+
+impl Frame for ServerConfirmsAuth {
+    const TAG: FrameType = FrameType::ServerConfirmsAuth;
+}
+
+impl Frame for ServerDeniesAuth {
+    const TAG: FrameType = FrameType::ServerDeniesAuth;
+}
+
+#[common_fields({
+    backtrace: Option<Backtrace>,
+    #[snafu(implicit)]
+    span_trace: n0_snafu::SpanTrace,
+})]
+#[allow(missing_docs)]
+#[derive(Debug, Snafu)]
+#[non_exhaustive]
+pub enum Error {
+    #[snafu(transparent)]
+    Websocket {
+        #[cfg(not(wasm_browser))]
+        source: tokio_websockets::Error,
+        #[cfg(wasm_browser)]
+        source: ws_stream_wasm::WsErr,
+    },
+    #[snafu(display("Handshake timeout reached"))]
+    Timeout { source: Elapsed },
+    #[snafu(display("Handshake stream ended prematurely"))]
+    UnexpectedEnd {},
+    #[snafu(transparent)]
+    FrameTypeError { source: FrameTypeError },
+    #[snafu(display("The relay denied our authentication ({reason})"))]
+    ServerDeniedAuth { reason: String },
+    #[snafu(display("Unexpected tag, got {frame_type}, but expected one of {expected_types:?}"))]
+    UnexpectedFrameType {
+        frame_type: FrameType,
+        expected_types: Vec<FrameType>,
+    },
+    #[snafu(display("Handshake failed while deserializing {frame_type} frame"))]
+    DeserializationError {
+        frame_type: FrameType,
+        source: postcard::Error,
+    },
+    #[cfg(feature = "server")]
+    /// Failed to deserialize client auth header
+    ClientAuthHeaderInvalid { value: HeaderValue },
+}
+
+impl ServerChallenge {
+    /// Generates a new challenge.
+    #[cfg(feature = "server")]
+    pub(crate) fn new(mut rng: impl RngCore + CryptoRng) -> Self {
+        let mut challenge = [0u8; 16];
+        rng.fill_bytes(&mut challenge);
+        Self { challenge }
+    }
+
+    /// The actual message bytes to sign (and verify against) for this challenge.
+    fn message_to_sign(&self) -> [u8; 32] {
+        blake3::derive_key(
+            "iroh-relay handshake v1 challenge signature",
+            &self.challenge,
+        )
+    }
+}
+
+impl ClientAuth {
+    /// Generates a signature for given challenge from the server.
+    pub(crate) fn new(secret_key: &SecretKey, challenge: &ServerChallenge) -> Self {
+        Self {
+            public_key: secret_key.public(),
+            signature: secret_key.sign(&challenge.message_to_sign()).to_bytes(),
+        }
+    }
+
+    /// Verifies this client's authentication given the challenge this was sent in response to.
+    #[cfg(feature = "server")]
+    pub(crate) fn verify(&self, challenge: &ServerChallenge) -> bool {
```

Comment:
I think using a `Result` is a bit cleaner, and matches the style of verification we have already